### PR TITLE
Add first-class Windows ROCm 7.14 support

### LINKED ISSUE #88: Feature request: first-class AMD ROCm 7.14 support on Windows
## Summary

I have already been using a locally modified version of `anima_lora` on an AMD Radeon GPU with ROCm, and the core training path works much better than I initially expected.

This is therefore not only a theoretical compatibility request: **the core LoRA training path, backward, optimizer step, and `torch.compile` have all worked locally on ROCm**. The remaining work appears to be mostly installation/dependency/backend selection rather than a large model-code port.

I think it would be practical to add a first-class AMD path that automatically installs/selects **ROCm 7.14** while preserving the current CUDA path for NVIDIA users.

## Tested environment

- OS: Windows 11
- GPU: AMD Radeon RX 9070 XT (`gfx1201`)
- Backend: ROCm 7.14 PyTorch environment
- Anima LoRA: locally modified from this repository

I have also been using AMD ROCm PyTorch with the same general Anima/ComfyUI stack, so the basic model operators are already known to work on this GPU family.

### 1. Core training

The normal training path works locally on ROCm, including:

- model loading
- forward pass
- loss computation
- backward pass
- AdamW optimizer step
- LoRA training

The project already relies mostly on normal PyTorch device APIs (`torch.cuda.*`, `device="cuda"`, etc.). ROCm PyTorch intentionally exposes the CUDA-compatible PyTorch API, so most of this code does not need an AMD-specific rewrite.

### 2. PyTorch SDPA works well

For AMD, `attn_mode = "torch"` is the best default in my testing.

The existing `networks/attention_dispatch.py` already has a clean PyTorch SDPA path using:

```python
torch.nn.functional.scaled_dot_product_attention(...)
```

so no new attention implementation is required just to support ROCm.

### 3. `torch.compile` works and should stay enabled

This is important: **I tested `torch.compile` on ROCm and it works in training.**

It also provides a meaningful memory reduction in my local tests, so I do **not** recommend disabling it for the AMD preset/backend.

Suggested AMD default:

```toml
attn_mode = "torch"
torch_compile = true
```

The project currently treats `torch.compile` as a major part of its VRAM/performance strategy, and ROCm does not require falling back to eager mode here in my tested setup.

## Flash Attention on AMD

I also tried an AMD-compatible Flash Attention implementation.

It worked, but for this Anima training workload it was approximately **3x slower than PyTorch SDPA** on my system.

Because of that, I would not make Flash Attention an AMD dependency or default. It only adds installation complexity while performing substantially worse in this case.

My recommendation is:

```text
NVIDIA -> flash (current default)
AMD    -> torch SDPA
```

AMD users could still manually experiment with other attention backends later, but the official ROCm path should stay simple and use SDPA by default.

## Proposed backend selection

Something like this would keep the current behavior while allowing an explicit override:

```text
ANIMA_BACKEND=auto   # default
ANIMA_BACKEND=cuda
ANIMA_BACKEND=rocm
```

Possible `auto` behavior:

```text
NVIDIA GPU -> CUDA path
AMD GPU    -> ROCm path
```

For unusual mixed-vendor systems, an explicit environment variable can override detection.

## Windows installer changes

The current `install.ps1` is strongly CUDA-specific: it checks for CUDA 13.2 and can download/install the CUDA toolkit automatically.

For a ROCm backend, I suggest branching before that step:

```powershell
if ($Backend -eq "cuda") {
    # existing CUDA 13.2 path
}
elseif ($Backend -eq "rocm") {
    # skip CUDA toolkit installation
    # install/use ROCm 7.14 PyTorch packages
}
```

The AMD path should automatically use the official ROCm 7.14 Windows PyTorch package source rather than requiring users to manually replace Torch after installation.

## `pyproject.toml` / `uv` is probably the main packaging issue

This looks like the most important part to solve correctly.

At the moment the project contains NVIDIA-specific dependencies/settings in the normal environment, including:

- CUDA 13.2 PyTorch index
- CUDA-specific prebuilt `flash-attn` wheels
- Windows Triton/CUDA-related dependencies

Simply installing ROCm Torch first is not enough if a later `uv sync` resolves the project back to the CUDA Torch packages.

So I think the clean implementation is to separate common dependencies from backend-specific dependencies/resolution:

```text
common
  - accelerate
  - transformers
  - diffusers
  - safetensors
  - etc.

cuda
  - CUDA PyTorch build
  - flash-attn CUDA wheel
  - CUDA/Triton-specific dependencies

rocm
  - ROCm 7.14 PyTorch build
  - no flash-attn by default
```

The important requirement is that later `uv run` / `uv sync` operations must not silently replace the ROCm PyTorch installation with the CUDA build.

## Suggested AMD defaults

I would start with only these backend-specific behavioral differences:

```toml
# AMD / ROCm
attn_mode = "torch"
torch_compile = true
gradient_checkpointing = false  # unchanged unless the selected preset needs it
optimizer_type = "AdamW"
```

Most of the rest of the existing configuration can remain shared with NVIDIA.

## GUI GPU monitor

`gui/gpu_status.py` currently uses `nvidia-smi`.

This does not appear to be a blocker because the widget already hides itself if `nvidia-smi` is unavailable. AMD monitoring can therefore be added later without blocking initial ROCm support.

## Scope I would recommend for an initial ROCm implementation

1. Add backend selection (`auto` / `cuda` / `rocm`).
2. Detect AMD vs NVIDIA for the default backend.
3. Keep the current CUDA 13.2 installation path unchanged for NVIDIA.
4. Add a ROCm 7.14 installation path for AMD.
5. Prevent CUDA-specific Torch / Flash Attention packages from being installed in the ROCm environment.
6. Default to `attn_mode = "torch"` on AMD.
7. Keep `torch_compile = true` on AMD.
8. Add a small post-install smoke test, for example:

```python
import torch

print("torch:", torch.__version__)
print("HIP:", torch.version.hip)
print("GPU:", torch.cuda.get_device_name(0))

assert torch.cuda.is_available()
assert torch.version.hip is not None
```

9. Treat AMD GPU monitoring / optional alternative attention backends as follow-up work rather than blockers.

## Why I think this is worth supporting

The biggest compatibility concerns I expected were attention backward and `torch.compile`, but in actual local use the important training path already works on ROCm.

The remaining gap is mostly making installation reproducible and preventing the CUDA-specific dependency setup from taking over the environment.

So this looks like a relatively small first-class backend addition compared with a full port.

I can also provide additional logs/benchmarks from the working ROCm setup if useful. I have already compared AMD Flash Attention against PyTorch SDPA and confirmed that SDPA is the preferable default on this workload.

### Current local conclusions

| Item | ROCm result |
|---|---|
| Core LoRA training | Works |
| Forward | Works |
| Backward | Works |
| AdamW step | Works |
| PyTorch SDPA | Works / preferred |
| AMD Flash Attention | Works, but ~3x slower than SDPA in my test |
| `torch.compile` | Works |
| `torch.compile` memory benefit | Confirmed |
| Recommended AMD attention default | `torch` |
| Recommended AMD compile default | enabled |

I have not exhaustively validated every optional training/inference feature in the repository, so I would initially describe this as support for the **core Anima LoRA training path on ROCm 7.14**, with optional methods/features validated incrementally.

### PR DESCRIPTION
**Pull request by qweqweewqe7-create**
Closes #88.

## What changed

- add `ANIMA_BACKEND=auto|cuda|rocm` selection to the Windows installer
- preserve the existing NVIDIA CUDA 13.2 path and skip the CUDA toolkit step for ROCm
- split Windows accelerator dependencies into mutually exclusive `cuda-windows` and `rocm-windows` extras
- resolve ROCm PyTorch 2.12 / torchvision 0.27 from AMD's official ROCm 7.14 multi-arch index
- scope the initial ROCm package set to RDNA 4 (`gfx1200` / `gfx1201`) instead of AMD's currently unresolvable `device-all` extra
- keep the selected backend across `make update`
- fall back from CUDA Flash Attention to PyTorch SDPA when HIP is detected, while leaving `torch_compile = true`
- add a post-install ROCm tensor + `torch.compile` + bf16 SDPA + backward smoke test
- document automatic selection, manual overrides, and manual `uv sync` commands

## Why

The previous Windows environment was CUDA-only at every layer: the CUDA toolkit check was unconditional, Torch came from the cu132 index, and Flash Attention / Triton were installed as normal Windows dependencies. Installing ROCm Torch first could therefore be undone by a later `uv sync`.

The backend extras and uv conflict declaration make CUDA and ROCm mutually exclusive. The installer and updater always reuse the chosen extra, so they cannot silently cross package indexes.

## AMD Flash Attention

This PR deliberately does not install an AMD Flash Attention implementation. On the tested Anima training workload, AMD FA worked but was about 3x slower than PyTorch SDPA. HIP therefore normalizes `attn_mode = "flash"` to `"torch"`; users keep the working compile path without adding an optional attention dependency.

## Validation

- `uv lock --check` — resolved 236 packages
- clean exact `uv sync --locked --no-dev --extra rocm-windows` — installed `torch==2.12.0+rocm7.14.0`, HIP `7.14.60850`, gfx1200/gfx1201 ROCm packages, and no CUDA Torch or Flash Attention
- `uv sync --dry-run --no-dev --extra cuda-windows` — retains `torch==2.12.0+cu132`, the existing Flash Attention wheel, and no ROCm packages
- `pytest tests/test_rocm_backend.py -q` — 9 passed
- PowerShell parser check for `install.ps1` — passed
- targeted Ruff checks — passed
- real AMD ROCm 7.14 smoke on RX 9060 XT (`gfx1200`) — tensor allocation, `torch.compile`, bf16 SDPA, backward, finite-value checks, and synchronize passed
- real AMD ROCm 7.14 smoke on RX 9070 XT (`gfx1201`) — the same checks passed

## Compatibility

- NVIDIA behavior remains CUDA 13.2 + Flash Attention.
- ROCm changes only backend selection and attention dispatch; optimizer, gradient checkpointing, and compile defaults are unchanged.
- Windows manual clones must select one extra for exact sync: `uv sync --extra cuda-windows` or `uv sync --extra rocm-windows`.

### CODE REVIEW

**Comment by sorryhyun** [comment]
@claude review pr

**Comment by qweqweewqe7-create** [comment]
Thanks for the review — fixed the false ROCm warning in `af75542a`.

The call sites now use a dedicated `needs_rocm_attention_fallback()` predicate, so `attn_mode=None`, explicit `torch`, and CUDA `flash` no longer report a ROCm fallback. Only an explicit `flash` request on a HIP build does. I also added regression coverage for the `None` case and extended the hardware smoke test to exercise bf16 PyTorch SDPA.

Fresh validation on this PC with an isolated, exact locked install:

- `torch==2.12.0+rocm7.14.0` / HIP `7.14.60850`
- RX 9060 XT (`gfx1200`): tensor + `torch.compile` + bf16 SDPA + backward passed
- RX 9070 XT (`gfx1201`): tensor + `torch.compile` + bf16 SDPA + backward passed
- `pytest tests/test_rocm_backend.py -q`: 9 passed
- Ruff on the touched Python files: passed

On the possible support scope: AMD's current ROCm 7.14 PyTorch install matrix also publishes `device-gfx1100`, `device-gfx1101`, `device-gfx1102`, and `device-gfx1103` extras, so the same packaging approach looks extendable to RDNA 3. I kept this PR's declared support at `gfx1200`/`gfx1201` because those are the architectures I could test on real hardware; RDNA 3 would still need a hardware smoke test before I would call it verified.

I also confirmed the review's index trade-off in the clean sync: with `unsafe-best-match`, AMD's index can supply packages such as NumPy/Pillow. The index currently needs to remain visible for the transitive ROCm device packages under uv, so I left that behavior unchanged rather than mixing it into this warning fix.

**Comment by sorryhyun** [comment]
nice. And could you tell me which devices are compatible, and is it binded to torch version (2.12) ? Finally dependency installation workflow for beginners (or scenario) and maintainability should be resolved so I can take cuda part without much care.

**Comment by qweqweewqe7-create** [comment]
Yes — for beginners, the intended Windows workflow is still the existing one-liner installer only:

```powershell
irm https://github.com/sorryhyun/anima_lora/releases/latest/download/install.ps1 | iex
```

The idea is that users should not need to know about `uv` extras, CUDA vs ROCm package indexes, or Torch wheel details. `install.ps1` detects the GPU vendor, chooses the backend-specific dependency set automatically, installs it, persists the backend choice, creates the desktop GUI shortcut, and launches the GUI. For updates, the normal beginner workflow is just the **Update** button in the GUI; the saved backend is reused automatically.

So the user-facing flow is effectively:

- NVIDIA -> CUDA 13.2 + Torch CUDA + Flash Attention
- AMD -> ROCm + Torch ROCm + PyTorch SDPA
- dependency install -> runtime verification -> desktop shortcut -> GUI launch

The manual `uv sync --extra cuda-windows` / `rocm-windows` commands are mainly for advanced/manual-clone use and do not need to be part of the normal beginner instructions.

For device compatibility, this PR currently *packages and verifies* only `gfx1200` / `gfx1201` (RDNA 4), because those are the architectures I can test on real hardware. ROCm 7.14 also publishes the corresponding device packages for RDNA 3 (`gfx1100` / `gfx1101` / `gfx1102` / `gfx1103`), so the same approach looks extendable there, but I would keep RDNA 3 labeled unverified until someone runs the same smoke test on real hardware.

One UX detail I think is worth tightening is AMD auto-detection: right now vendor detection can select ROCm for any AMD GPU, while the packaged set in this PR is only `gfx1200`/`gfx1201`. We should either reject unsupported/unverified AMD hardware early with a clear message, or expand the device extras after hardware validation, rather than letting a beginner discover incompatibility after a large install.

Regarding Torch 2.12: the current PR is intentionally pinned to `torch==2.12.0+rocm7.14.0` / `torchvision==0.27.0+rocm7.14.0` because that is the ROCm 7.14 Windows package combination being integrated here and it also matches anima_lora's current Torch 2.12 CUDA baseline. The backend/runtime logic itself is not fundamentally tied to Torch 2.12.

For additional confidence on that point, in my separate local ROCm environments I have also run the same core Anima/ROCm path successfully with newer development stacks:

- Torch 2.14 alpha + ROCm 7.15 alpha
- Torch 2.15 alpha + ROCm 10.1 alpha

Those are not the locked/official dependency set proposed by this PR, so I would not advertise them as supported versions, but they suggest the implementation is not intrinsically coupled to 2.12. When the project moves its Torch baseline later, the ROCm pin should be able to move with it rather than requiring a backend redesign.

My goal is that adding ROCm should not create ongoing CUDA maintenance burden: CUDA users keep the same one-line installer, GUI update workflow, CUDA toolkit path, Flash Attention path, and pinned CUDA dependency set; ROCm-specific dependency choices stay isolated behind the backend selection.

**Comment by sorryhyun** [comment]
I see. Let's add dedicated doc for rocm and add brief note about rocm support in guidelines, then flip pr to be ready
